packages/webpack/src/runtime/runtimeBuilder.js

```diff
@@ -1,11 +1,55 @@
 const { RUNTIME_KEY } = require('../ENUM.json')
 const diag = require('../buildtime/diagnostics.js')
-const { assembleRuntime } = require('./assemble.js')
+const { assembleRuntime, prepareSource } = require('./assemble.js')
 const path = require('node:path')
 
-/** @import {LavaMoatPluginOptions} from '../buildtime/types' */
+const { RuntimeModule } = require('webpack')
+
+class VirtualRuntimeModule extends RuntimeModule {
+  /**
+   * @param {Object} options - The options for the VirtualRuntimeModule.
+   * @param {string} options.name - The name of the module.
+   * @param {string} options.source - The source code of the module.
+   * @param {number} [options.stage] - The stage of runtime. One of
+   *   RuntimeModule.STAGE_*.
+   * @param {boolean} [options.withoutClosure] - Make the source code run
+   *   outside the closure for a runtime module
+   */
+  constructor({
+    name,
+    source,
+    stage = RuntimeModule.STAGE_NORMAL,
+    withoutClosure = false,
+  }) {
+    super(name, stage)
+    this.withoutClosure = withoutClosure
+    this.virtualSource = `;${source};`
+  }
+  shouldIsolate() {
+    return !this.withoutClosure
+  }
+
+  generate() {
+    return this.virtualSource
+  }
+}
+
+/** @import {LavaMoatPluginOptions, LavaMoatChunkRuntimeConfiguration} from '../buildtime/types' */
 /** @import {LavaMoatPolicy} from 'lavamoat-core' */
-/** @import {RuntimeModule} from './assemble.js' */
+/** @import {RuntimeFragment} from './assemble.js' */
+/** @import {Chunk} from 'webpack' */
+
+/**
+ * @typedef {Object} LavaMoatRuntimeIdentifiers
+ * @property {string} root - Root identifier
+ * @property {[string, (string | number)[]][]} identifiersForModuleIds - Module
```

Comment:
if this is a data structure we own, maybe we should change it.  at the very least, since it's a tuple, we can give the members names, e.g.:

```ts
type Thing = [foo: string, bar: (string|number)[]][]
```